File: trunk/net/NetTupleEngine/Compound.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace it.mintlab.desktopnet.tupleengine
{
    public class Compound : Term, IRefObject
    {
        private string rootValue;
        private List<Term> leafValues;
        private object refObject;
        private bool withRefObject;

        public Compound(string rootValue, List<Term> leafValues)
        {
            this.rootValue = rootValue;
            this.leafValues = leafValues;
            this.withRefObject = false;
        }

        public void setRefObject(object obj)
        {
            refObject = obj;
            withRefObject = true;
        }

        public override string ToString()
        {
            string leafs = "";
            foreach (Term leaf in leafValues)
            {
                leafs += leaf.ToString() + ",";
            }
            return rootValue + "(" + leafs.Remove(leafs.Length - 1) + ")";
        }

        #region Term Membri di

        internal override bool unify(Term t, List<Var> vars1, List<Var> vars2)
        {
            if (t is Compound)
            {
                Compound tc = (Compound)t;
                if (this.rootValue.Equals(tc.rootValue) &&
                    this.leafValues.Count == tc.leafValues.Count)
                {
                    for (int i = 0; i < leafValues.Count; i++)
                    {
                        if (!this.leafValues[i].unify(tc.leafValues[i], vars1, vars2)) return false;
                    }
                    return true;
                }
            }
            if (t is Var)
            {
                return t.unify(this, vars2, vars1);
            }
            return false;
        }

        public override object getRefObject()
        {
            if (withRefObject)
                return refObject;
            else
                if (rootValue.StartsWith("'"))
                    return rootValue.Replace("'", "");
                else if (rootValue.StartsWith("\""))
                    return rootValue.Replace("\"", "");
                else
                    return rootValue;
        }

        #endregion
    }
}

File: java/JavaTupleEngine/src/it/mintlab/desktopjava/tupleengine/Compound.java
package it.mintlab.desktopjava.tupleengine;

	public class Compound extends Term implements IRefObject
	{
		private String rootValue;
		private java.util.ArrayList<Term> leafValues;
		private Object refObject;
		private boolean withRefObject;

		public Compound(String rootValue, java.util.ArrayList<Term> leafValues)
		{
			this.rootValue = rootValue;
			this.leafValues = leafValues;
			this.withRefObject = false;
		}

		public final void setRefObject(Object obj)
		{
			refObject = obj;
			withRefObject = true;
		}

		@Override
		public String toString()
		{
			String leafs = "";
			for (Term leaf : leafValues)
			{
				leafs += leaf.toString() + ",";
			}
			return rootValue + "(" + leafs.substring(0, leafs.length() - 1) + ")";
		}

		@Override
		public boolean unify(Term t, java.util.ArrayList<Var> vars1, java.util.ArrayList<Var> vars2)
		{
			if (t instanceof Compound)
			{
				Compound tc = (Compound)t;
				if (this.rootValue.equals(tc.rootValue) && this.leafValues.size() == tc.leafValues.size())
				{
					for (int i = 0; i < leafValues.size(); i++)
					{
						if (!this.leafValues.get(i).unify(tc.leafValues.get(i), vars1, vars2))
						{
							return false;
						}
					}
					return true;
				}
			}
			if (t instanceof Var)
			{
				return t.unify(this, vars2, vars1);
			}
			return false;
		}

		@Override
		public Object getRefObject()
		{
			return withRefObject ? refObject : rootValue;
		}
	}